source/gloperate/include/gloperate/pipeline/Input.inl

```diff
@@ -26,6 +29,29 @@ Input<T>::~Input()
 template <typename T>
 void Input<T>::onValueChanged(const T & value)
 {
+    // Get current thread ID
+    std::thread::id this_id = std::this_thread::get_id();
+
+    // Initialize guard for the current thread
+    if (this->m_cycleGuard.find(this_id) == this->m_cycleGuard.end())
```

Comment:
I think we need mutexes here for all the operations (searching, inserting and updating the map).
This is also required for the Output class.